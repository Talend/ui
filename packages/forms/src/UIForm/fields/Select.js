import React, { PropTypes } from 'react';
import classNames from 'classnames';

import Message from '../Message';

function getSelectedOptions(select, multiple) {
	if (multiple) {
		return Array.from(select.options)
			.filter(option => option.selected)
			.map(option => option.value);
	}

	return select.value;
}

export default function Select(props) {
	const { id, isValid, errorMessage, onChange, schema, value } = props;
	const { autoFocus, description, disabled, placeholder, readOnly, title } = schema;

	const groupsClassNames = classNames(
		'form-group',
		{ 'has-error': !isValid },
	);
	const multiple = schema.schema.type === 'array' && schema.schema.uniqueItems;

	return (
		<div className={groupsClassNames}>
			<select
				id={id}
				multiple={multiple}
				autoFocus={autoFocus}
				className="form-control"
				disabled={disabled}
				onChange={
					event => onChange(event, schema, getSelectedOptions(event.target, multiple))
				}
				readOnly={readOnly}
				value={value}
			>
				<option disabled>{placeholder}</option>
				{
					schema.titleMap && schema.titleMap.map((option, index) => {
						const optionProps = {
							key: index,
							value: option.value,
						};
						return (
							<option
								{...optionProps}
							>
								{option.name}
							</option>
						);
					})
				}
			</select>
			<label htmlFor={id} className="control-label">{title}</label>
			<Message
				errorMessage={errorMessage}
				description={description}
				isValid={isValid}
			/>
		</div>
	);
}

if (process.env.NODE_ENV !== 'production') {
	Select.propTypes = {
		id: PropTypes.string,
		isValid: PropTypes.bool,
		errorMessage: PropTypes.string,
		onChange: PropTypes.func,
		schema: PropTypes.shape({
			autoFocus: PropTypes.bool,
			description: PropTypes.string,
			disabled: PropTypes.bool,
			placeholder: PropTypes.string,
			readOnly: PropTypes.bool,
			title: PropTypes.string,
			type: PropTypes.string,
		}),
		value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.array]),
	};
}
Select.defaultProps = {
	isValid: true,
	value: '',
};
