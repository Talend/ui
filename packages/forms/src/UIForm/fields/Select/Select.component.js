import PropTypes from 'prop-types';
import React from 'react';
import FieldTemplate from '../FieldTemplate';
import { generateDescriptionId, generateErrorId } from '../../Message/generateId';

function getSelectedOptions(select, multiple) {
	if (multiple) {
		return Array.prototype.slice
			.call(select.options)
			.filter(option => option.selected)
			.map(option => option.value);
	}

	return select.value;
}

export default function Select({
	id,
	isValid,
	errorMessage,
	onChange,
	onFinish,
	schema,
	value,
	valueIsUpdating,
}) {
	const { autoFocus, description, disabled = false, placeholder, readOnly = false, title } = schema;
	const descriptionId = generateDescriptionId(id);
	const errorId = generateErrorId(id);

	const multiple = schema.schema.type === 'array' && schema.schema.uniqueItems;

	return (
		<FieldTemplate
			description={description}
			errorMessage={errorMessage}
			descriptionId={descriptionId}
			errorId={errorId}
			id={id}
			isValid={isValid}
			label={title}
			labelAfter
			required={schema.required}
			valueIsUpdating={valueIsUpdating}
		>
			<select
				id={id}
				multiple={multiple}
				autoFocus={autoFocus}
				className="form-control"
				disabled={disabled || valueIsUpdating}
				onChange={event => {
					const payload = { schema, value: getSelectedOptions(event.target, multiple) };
					onChange(event, payload);
					onFinish(event, payload);
				}}
				readOnly={readOnly}
				value={value}
				// eslint-disable-next-line jsx-a11y/aria-proptypes
				aria-invalid={!isValid}
				aria-required={schema.required}
				aria-describedby={`${descriptionId} ${errorId}`}
			>
				{placeholder ? (
					<option disabled value="">
						{placeholder}
					</option>
				) : null}
				{schema.titleMap &&
					schema.titleMap.map((option, index) => {
						const optionProps = {
							key: index,
							value: option.value,
						};
						return <option {...optionProps}>{option.name}</option>;
					})}
			</select>
		</FieldTemplate>
	);
}

if (process.env.NODE_ENV !== 'production') {
	Select.propTypes = {
		id: PropTypes.string,
		isValid: PropTypes.bool,
		errorMessage: PropTypes.string,
		onChange: PropTypes.func.isRequired,
		onFinish: PropTypes.func.isRequired,
		schema: PropTypes.shape({
			autoFocus: PropTypes.bool,
			description: PropTypes.string,
			disabled: PropTypes.bool,
			placeholder: PropTypes.string,
			readOnly: PropTypes.bool,
			title: PropTypes.string,
			titleMap: PropTypes.arrayOf(
				PropTypes.shape({
					name: PropTypes.string.isRequired,
					value: PropTypes.string.isRequired,
				}),
			),
			type: PropTypes.string,
		}),
		value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.array]),
		valueIsUpdating: PropTypes.bool,
	};
}

Select.defaultProps = {
	isValid: true,
	schema: {},
	value: '',
};
