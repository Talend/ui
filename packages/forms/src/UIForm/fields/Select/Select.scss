.select {
	:global(.Select-control) {
		border: none;
		border-bottom: 1px solid $black;
		border-radius: 0;
	}
	:global(.is-focused):not(:global(.is-open)) > :global(.Select-control) {
		border-color: none;
	}
}
