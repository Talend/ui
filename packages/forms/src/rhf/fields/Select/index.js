import RHFSelect from './RHFSelect.component';

export default RHFSelect;
