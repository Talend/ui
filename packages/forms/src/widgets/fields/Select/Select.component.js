import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import FieldTemplate from '../../templates/FieldTemplate';
import { generateDescriptionId, generateErrorId } from '../../templates/utils';

const Select = React.forwardRef((props, ref) => {
	const { className, description, error, label, inProgress, placeholder, ...rest } = props;
	const { id, required, options } = rest;

	const descriptionId = generateDescriptionId(id);
	const errorId = generateErrorId(id);

	return (
		<FieldTemplate
			description={description}
			descriptionId={descriptionId}
			error={error}
			errorId={errorId}
			id={id}
			label={label}
			inProgress={inProgress}
			required={required}
		>
			<select
				{...rest}
				ref={ref}
				className={classnames('form-control', className)}
				// eslint-disable-next-line jsx-a11y/aria-proptypes
				aria-invalid={!!error}
				aria-required={required}
				aria-describedby={`${descriptionId} ${errorId}`}
			>
				{placeholder ? (
					<option disabled value="">
						{placeholder}
					</option>
				) : null}
				{options &&
					options.map(option => {
						const optionProps = {
							key: option.value,
							value: option.value,
						};
						return <option {...optionProps}>{option.name}</option>;
					})}
			</select>
		</FieldTemplate>
	);
});

if (process.env.NODE_ENV !== 'production') {
	Select.propTypes = {
		id: PropTypes.string.isRequired,
		className: PropTypes.string,
		description: PropTypes.string,
		error: PropTypes.string,
		label: PropTypes.string.isRequired,
		inProgress: PropTypes.bool,
		rules: PropTypes.object,
		placeholder: PropTypes.string,
	};
}

export default Select;
