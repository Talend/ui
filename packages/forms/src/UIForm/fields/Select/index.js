import Select from './Select.component';

export default Select;
