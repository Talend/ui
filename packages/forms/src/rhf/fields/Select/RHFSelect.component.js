import React from 'react';
import PropTypes from 'prop-types';
import { useFormContext } from 'react-hook-form';
import get from 'lodash/get';

import Select from '../../../widgets/fields/Select';

function RHFSelect(props) {
	const { rules = {}, ...rest } = props;
	const { errors, register } = useFormContext();
	const error = get(errors, rest.name)?.message;
	return <Select {...rest} ref={register(rules)} error={error} />;
}

if (process.env.NODE_ENV !== 'production') {
	RHFSelect.propTypes = {
		rules: PropTypes.object,
	};
}

export default RHFSelect;
