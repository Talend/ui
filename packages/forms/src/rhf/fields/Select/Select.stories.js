/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { useForm, FormProvider } from 'react-hook-form';
import { action } from '@storybook/addon-actions';
import Select from '.';

export default {
	title: 'RHF/Select',

	parameters: {
		component: Select,
	},
};

export const States = () => {
	const rhf = useForm();
	return (
		<FormProvider {...rhf}>
			<form onSubmit={rhf.handleSubmit(action('submit'))} noValidate>
				<Select
					id="name"
					name="default"
					label="Default"
					options={[
						{ value: 'blue', name: 'Blue color' },
						{ value: 'red', name: 'Red color' },
					]}
				/>
				<Select
					id="disabled"
					name="disabled"
					label="Disabled"
					options={[
						{ value: 'blue', name: 'Blue color' },
						{ value: 'red', name: 'Red color' },
					]}
					disabled
				/>
				<button type="submit" className="btn btn-primary">
					Submit
				</button>
			</form>
		</FormProvider>
	);
};

export const Description = () => {
	const rhf = useForm();

	return (
		<FormProvider {...rhf}>
			<form onSubmit={rhf.handleSubmit(action('submit'))} noValidate>
				<Select
					id="name"
					type="text"
					name="default"
					label="Default"
					options={[
						{ value: 'blue', name: 'Blue color' },
						{ value: 'red', name: 'Red color' },
					]}
					description="This field has a description"
				/>
				<button type="submit" className="btn btn-primary">
					Submit
				</button>
			</form>
		</FormProvider>
	);
};

export const Validation = () => {
	const rhf = useForm({ mode: 'onBlur' });

	return (
		<FormProvider {...rhf}>
			<form onSubmit={rhf.handleSubmit(action('submit'))} noValidate>
				<Select
					id="required"
					name="required"
					label="Required"
					placeholder="Choose an option"
					options={[
						{ value: 'blue', name: 'Blue color' },
						{ value: 'red', name: 'Red color' },
					]}
					rules={{ required: 'This is required' }}
					required
				/>

				<Select
					id="notBlue"
					name="notBlue"
					label="Not blue"
					options={[
						{ value: 'blue', name: 'Blue color' },
						{ value: 'red', name: 'Red color' },
					]}
					rules={{
						validate(value) {
							return value === 'blue' ? 'This should not be blue' : null;
						},
					}}
					required
				/>

				<button type="submit" className="btn btn-primary">
					Submit
				</button>
			</form>
		</FormProvider>
	);
};
