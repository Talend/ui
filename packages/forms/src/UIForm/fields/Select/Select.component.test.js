import React from 'react';
import { shallow } from 'enzyme';
import omit from 'lodash/omit';

import Select from './Select.component';

describe('Select field', () => {
	const schema = {
		autoFocus: true,
		description: 'Select me',
		placeholder: 'Please select a value',
		required: true,
		schema: {
			enum: ['foo', 'bar', 'lol'],
			type: 'string',
		},
		title: 'My Select title',
		titleMap: [
			{ name: 'My foo title', value: 'foo' },
			{ name: 'My bar title', value: 'bar' },
			{ name: 'My lol title', value: 'lol' },
		],
		type: 'select',
	};

	it('should render simple select', () => {
		// when
		const wrapper = shallow(
			<Select
				id={'mySelect'}
				isValid
				errorMessage={'My Error Message'}
				onChange={jest.fn()}
				onFinish={jest.fn()}
				schema={schema}
				value={'lol'}
			/>,
		);

		// then
		expect(wrapper.getElement()).toMatchSnapshot();
	});

	it('should render simple select without placeholder', () => {
		// given
		const localSchema = omit(schema, 'placeholder');

		// when
		const wrapper = shallow(
			<Select
				id={'mySelect'}
				isValid
				errorMessage={'My Error Message'}
				onChange={jest.fn()}
				onFinish={jest.fn()}
				schema={localSchema}
				value={'lol'}
			/>,
		);

		// then
		expect(wrapper.getElement()).toMatchSnapshot();
	});

	it('should render select multiple', () => {
		// given
		const multipleSchema = {
			...schema,
			multiple: true,
			schema: {
				type: 'array',
				uniqueItems: true,
				items: schema.schema,
			},
		};

		// when
		const wrapper = shallow(
			<Select
				id={'mySelect'}
				isValid
				errorMessage={'My Error Message'}
				onChange={jest.fn()}
				onFinish={jest.fn()}
				schema={multipleSchema}
				value={['foo', 'lol']}
			/>,
		);

		// then
		expect(wrapper.getElement()).toMatchSnapshot();
	});

	it('should render disabled input', () => {
		// given
		const disabledSchema = {
			...schema,
			disabled: true,
		};

		// when
		const wrapper = shallow(
			<Select
				id={'mySelect'}
				isValid
				errorMessage={'My Error Message'}
				onChange={jest.fn()}
				onFinish={jest.fn()}
				schema={disabledSchema}
				value={'lol'}
			/>,
		);

		// then
		expect(wrapper.getElement()).toMatchSnapshot();
	});

	it('should render readOnly input', () => {
		// given
		const readOnlySchema = {
			...schema,
			readOnly: true,
		};

		// when
		const wrapper = shallow(
			<Select
				id={'mySelect'}
				isValid
				errorMessage={'My Error Message'}
				onChange={jest.fn()}
				onFinish={jest.fn()}
				schema={readOnlySchema}
				value={'lol'}
			/>,
		);

		// then
		expect(wrapper.getElement()).toMatchSnapshot();
	});

	it('should trigger onChange', () => {
		// given
		const onChange = jest.fn();
		const wrapper = shallow(
			<Select
				id={'mySelect'}
				isValid
				errorMessage={'My Error Message'}
				onChange={onChange}
				onFinish={jest.fn()}
				schema={schema}
				value={'lol'}
			/>,
		);
		const event = { target: { value: 'bar' } };

		// when
		wrapper.find('select').simulate('change', event);

		// then
		expect(onChange).toBeCalledWith(event, { schema, value: 'bar' });
	});

	it('should trigger array onChange', () => {
		// given
		const onChange = jest.fn();
		const onFinish = jest.fn();
		const multipleSchema = {
			...schema,
			multiple: true,
			schema: {
				type: 'array',
				uniqueItems: true,
				items: schema.schema,
			},
		};
		const event = {
			target: {
				options: [
					{ value: 'foo', selected: true },
					{ value: 'bar', selected: false },
					{ value: 'lol', selected: true },
				],
			},
		};
		const wrapper = shallow(
			<Select
				id={'mySelect'}
				isValid
				errorMessage={'My Error Message'}
				onChange={onChange}
				onFinish={onFinish}
				schema={multipleSchema}
				value={['foo', 'lol']}
			/>,
		);

		// when
		wrapper.find('select').simulate('change', event);

		// then
		expect(onChange).toBeCalledWith(event, { schema: multipleSchema, value: ['foo', 'lol'] });
		expect(onFinish).toBeCalledWith(event, { schema: multipleSchema, value: ['foo', 'lol'] });
	});
});
