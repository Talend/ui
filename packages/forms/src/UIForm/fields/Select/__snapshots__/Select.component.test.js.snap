// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select field should render disabled input 1`] = `
<FieldTemplate
  description="Select me"
  descriptionId="mySelect-description"
  errorId="mySelect-error"
  errorMessage="My Error Message"
  id="mySelect"
  isValid={true}
  label="My Select title"
  labelAfter={true}
  required={true}
>
  <select
    aria-describedby="mySelect-description mySelect-error"
    aria-invalid={false}
    aria-required={true}
    autoFocus={true}
    className="form-control"
    disabled={true}
    id="mySelect"
    multiple={false}
    onChange={[Function]}
    readOnly={false}
    value="lol"
  >
    <option
      disabled={true}
      value=""
    >
      Please select a value
    </option>
    <option
      value="foo"
    >
      My foo title
    </option>
    <option
      value="bar"
    >
      My bar title
    </option>
    <option
      value="lol"
    >
      My lol title
    </option>
  </select>
</FieldTemplate>
`;

exports[`Select field should render readOnly input 1`] = `
<FieldTemplate
  description="Select me"
  descriptionId="mySelect-description"
  errorId="mySelect-error"
  errorMessage="My Error Message"
  id="mySelect"
  isValid={true}
  label="My Select title"
  labelAfter={true}
  required={true}
>
  <select
    aria-describedby="mySelect-description mySelect-error"
    aria-invalid={false}
    aria-required={true}
    autoFocus={true}
    className="form-control"
    id="mySelect"
    multiple={false}
    onChange={[Function]}
    readOnly={true}
    value="lol"
  >
    <option
      disabled={true}
      value=""
    >
      Please select a value
    </option>
    <option
      value="foo"
    >
      My foo title
    </option>
    <option
      value="bar"
    >
      My bar title
    </option>
    <option
      value="lol"
    >
      My lol title
    </option>
  </select>
</FieldTemplate>
`;

exports[`Select field should render select multiple 1`] = `
<FieldTemplate
  description="Select me"
  descriptionId="mySelect-description"
  errorId="mySelect-error"
  errorMessage="My Error Message"
  id="mySelect"
  isValid={true}
  label="My Select title"
  labelAfter={true}
  required={true}
>
  <select
    aria-describedby="mySelect-description mySelect-error"
    aria-invalid={false}
    aria-required={true}
    autoFocus={true}
    className="form-control"
    id="mySelect"
    multiple={true}
    onChange={[Function]}
    readOnly={false}
    value={
      Array [
        "foo",
        "lol",
      ]
    }
  >
    <option
      disabled={true}
      value=""
    >
      Please select a value
    </option>
    <option
      value="foo"
    >
      My foo title
    </option>
    <option
      value="bar"
    >
      My bar title
    </option>
    <option
      value="lol"
    >
      My lol title
    </option>
  </select>
</FieldTemplate>
`;

exports[`Select field should render simple select 1`] = `
<FieldTemplate
  description="Select me"
  descriptionId="mySelect-description"
  errorId="mySelect-error"
  errorMessage="My Error Message"
  id="mySelect"
  isValid={true}
  label="My Select title"
  labelAfter={true}
  required={true}
>
  <select
    aria-describedby="mySelect-description mySelect-error"
    aria-invalid={false}
    aria-required={true}
    autoFocus={true}
    className="form-control"
    id="mySelect"
    multiple={false}
    onChange={[Function]}
    readOnly={false}
    value="lol"
  >
    <option
      disabled={true}
      value=""
    >
      Please select a value
    </option>
    <option
      value="foo"
    >
      My foo title
    </option>
    <option
      value="bar"
    >
      My bar title
    </option>
    <option
      value="lol"
    >
      My lol title
    </option>
  </select>
</FieldTemplate>
`;

exports[`Select field should render simple select without placeholder 1`] = `
<FieldTemplate
  description="Select me"
  descriptionId="mySelect-description"
  errorId="mySelect-error"
  errorMessage="My Error Message"
  id="mySelect"
  isValid={true}
  label="My Select title"
  labelAfter={true}
  required={true}
>
  <select
    aria-describedby="mySelect-description mySelect-error"
    aria-invalid={false}
    aria-required={true}
    autoFocus={true}
    className="form-control"
    id="mySelect"
    multiple={false}
    onChange={[Function]}
    readOnly={false}
    value="lol"
  >
    <option
      disabled={true}
    />
    <option
      value="foo"
    >
      My foo title
    </option>
    <option
      value="bar"
    >
      My bar title
    </option>
    <option
      value="lol"
    >
      My lol title
    </option>
  </select>
</FieldTemplate>
`;
